package Thilenius.BFEProtos;

option java_package = "com.thilenius.blaze.frontend.protos";
option java_outer_classname = "BFEProtos";

message BFEMessage {
  extensions 100 to max;
}

/******************************************************************************/
/* Level Control                                                              */
/******************************************************************************/
message BFELoadLevelRequest {
  extend BFEMessage {
    optional BFELoadLevelRequest BFELoadLevelRequest_ext = 100;
  }

  required string auth_token = 1;
  required int32 levelNumber = 2;
  optional int32 seed = 3;
}

message BFELoadLevelResponse {
  extend BFEMessage {
    optional BFELoadLevelResponse BFELoadLevelResponse_ext = 101;
  }

  optional string failure_reason = 1;
  optional int32 spark_count = 2;
}

/******************************************************************************/
/* Spark Control                                                              */
/******************************************************************************/
message BFESparkCommand {
  extend BFEMessage {
    optional BFESparkCommand BFESparkCommand_ext = 102;
  }

  enum CommandType {
    MOVE_FORWARD     = 0;
    MOVE_BACKWARD    = 1;
    MOVE_UP          = 2;
    MOVE_DOWN        = 3;
    TURN_LEFT        = 4;
    TURN_RIGHT       = 5;
  }

  required CommandType command = 1;
  required int32 spark_id = 2;
  required string auth_token = 3;
  optional string pipline_token = 4;
}

message BFESparkResponse {
  extend BFEMessage {
    optional BFESparkResponse BFESparkResponse_ext = 103;
  }

  optional string pipline_token = 1;
  optional string failure_reason = 2;
  optional bool response_bool = 3;
  optional int32 response_int = 4;
  optional float response_float = 5;
  optional bytes response_bytes = 6;
}

/******************************************************************************/
/* Basic                                                                      */
/******************************************************************************/
message BFEInfoQueryRequest {
  extend BFEMessage {
    optional BFEInfoQueryRequest BFEInfoQueryRequest_ext = 104;
  }

  required string auth_token = 1;
}

message BFEInfoQueryResponse {
  extend BFEMessage {
    optional BFEInfoQueryResponse BFEInfoQueryResponse_ext = 105;
  }

  optional string failure_reason = 1;
  optional string blaze_response = 2;
}

/******************************************************************************/
/* Delta Compression                                                          */
/******************************************************************************/
message Repository {
  optional string repo_uuid = 1;

  repeated Revision revisions = 2;
}

message Revision {
  optional string comment = 1;
  optional string timestamp = 2;

  repeated FileDelta file_deltas = 3;
}

message FileDelta {
  enum ChangeType {
    CREATE = 1;
    MODIFY = 2;
    DELETE = 3;
  }

  required string path = 1;
  required ChangeType change_type = 2;

  repeated LexicalDelta lexical_deltas = 3;
}

message LexicalDelta {
  required int32 from_begin = 1;
  required int32 from_length = 2;
  required int32 to_begin = 3;
  required int32 to_length = 4;
  optional string text = 5;
}

/******************************************************************************/
/* Code Submit                                                                */
/******************************************************************************/
message BFETextFile {
  optional string name = 1;
  optional string extension = 2;
  optional string modify_date = 3;
  optional string contents = 4;
}

message BFECodeSubmitRequest {
  extend BFEMessage {
    optional BFECodeSubmitRequest BFECodeSubmitRequest_ext = 106;
  }

  required string auth_token = 1;
  repeated BFETextFile code_files = 2; 
}

message BFECodeSubmitResponse {
  extend BFEMessage {
    optional BFECodeSubmitResponse BFECodeSubmitResponse_ext = 107;
  }

  optional string failure_reason = 1;
}